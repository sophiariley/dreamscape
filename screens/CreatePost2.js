import React, {useState} from "react";
import {StyleSheet, Text, TextInput, View, KeyboardAvoidingView, TouchableOpacity, Pressable, Image} from "react-native";
import {Ionicons} from 'react-native-vector-icons';
import * as ImagePicker from 'expo-image-picker';
import { ref, uploadBytes, uploadBytesResumable } from "firebase/storage"
import { db, storage } from "../firebase-config";
import { collection, query, where, onSnapshot, getDocs, getDoc, getDocuments, doc, snapshotEqual, getCountFromServer } from "firebase/firestore";


const CreatePost2 = ({navigation, route}) => {

    const username = route.params.username;
    const userID = route.params.userID;

    const [caption, setCaption] = useState(' ');
    const [location, setLocation] = useState(' ');

    const [image, setImage] = useState(null);

    const metadata = {
        contentType: 'image/jpeg'
    };

    const pickImage = async () => {
        let result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.All,
            allowsEditing: true,
            aspect: [4, 3],
            quality: 1,
        });

        console.log(result);

        if (!result.canceled) { // cancelled
            setImage(result.assets[0].uri);
            //call method to add pic to storage and firestore here
            const uri = result.assets[0].uri;
            console.log("upload uri: ", uri);
            const picname = uri.substring(uri.lastIndexOf('/') + 1);
            const file = uri.substring(7);
            console.log('File: ', file);
            //const blob = await result.blob();
            uploadPic(file, picname);
        }
    };

    function uploadPic(image, name) {
        const storageRef = ref(storage, 'images/' + name);
        uploadBytes(storageRef, image, metadata);
    }

    return (
        <View style={styles.container}>
            
            <View style={styles.photoContainer}>
                {image && <Image source={{ uri: image }} style={{ width: 370, height: 320 }} />}
            </View>
            
            <View style={styles.chooseImageButtonContainer}>
                <Pressable onPress={pickImage}>
                    <Text style={styles.chooseImageText}>Choose image</Text>
                </Pressable>
            </View>
            
            <KeyboardAvoidingView style={styles.captionContainer}>
                <TextInput
                     style={styles.captionText} 
                     placeholder="Write caption..."
                     value={caption}
                     onChangeText={text => setCaption(text)}
                />
            </KeyboardAvoidingView>
            
            <View style={styles.location}>
                <Ionicons style={styles.globe} name='globe-outline' size={30} color='#3A6496'/>
                <TextInput 
                    style={styles.locationText}
                    placeholder= 'Add Location'
                    value={location}
                    placeholderTextColor="#3A6496"
                    onChangeText={text => setLocation(text)}
                />
            </View>
            
            <View style={styles.postButtonContainer}>
                <TouchableOpacity 
                    style={styles.postButton}
                    onPress={() => navigation.navigate('Home', {
                        username: username,
                        userID: userID,
                    })}>
                    <Text style={styles.postText}>Post</Text>
                </TouchableOpacity>
            
            </View>
        </View>
    )
}

export default CreatePost2;

const styles = StyleSheet.create({
    container: {
        backgroundColor: 'white',
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center', 
        flexDirection: 'column',
    },
    captionContainer: {
        backgroundColor: '#F6F6F6',
        opacity: .8,
        alignSelf: 'center',
        width: '90%',
        height: '20%',
        borderRadius: 5,
        bottom: 15
    },
    photoContainer: {
        backgroundColor: '#F6F6F6',
        backgroundOpacity: .8,
        alignSelf: 'center',
        width: '90%',
        height: '50%',
        borderRadius: 5,
        top: -50
    },
    captionText: {
        fontSize: 20,
        color: '#3A6496',
        opacity: .5,
        height: '20%',
        width: '80%',
        borderRadius: 5,
        marginTop: 5,
        marginBottom: 10,
        paddingLeft: 10
    },
    postButtonContainer: {
        width: '30%',
        height: '5%',
        flex: 1,
        alignSelf: 'flex-end',
        position: 'absolute',
        bottom: 40,
        right: 20
    },
    postButton: {
        backgroundColor: '#D28A8E', // pink color
        width: '100%',
        padding: 15,
        borderRadius: 5,
        alignItems: 'center',
        position: 'absolute'
    },
    postText: {
        fontSize: 20,
        color: '#F6F6F6'
    },
    location: {
        flexDirection: 'row',
        alignItems: 'center',
        paddingBottom: 30,
        paddingTop: 20,
        position: 'absolute',
        bottom: -10,
        alignSelf: "flex-start",
        left: 10
    },
    globe: {
        padding: 10,
    },
    locationText: {
        color: '#3A6496',
        fontSize: 20,
        padding: 5,
    },
    chooseImageText: {
        color:'#F6F6F6',
        alignSelf: 'center',
        fontSize: 20,
        paddingTop: 2
    },
    chooseImageButtonContainer: {
        backgroundColor: '#3A6496',
        width: '40%',
        height: '5%',
        alignSelf: 'center',
        borderRadius: 5,
        bottom: 30
    }
})