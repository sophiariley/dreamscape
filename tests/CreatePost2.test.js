import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react-native'; 
import CreatePost2 from './CreatePost2';
import {create} from 'react-test-renderer';

// basic rendering test
test('renders correctly across screens', () => {
  const tree = create(<CreatePost2 />);
  expect(tree).toMatchSnapshot();
});


// Testing "choose image" button
describe('Testing choose image button', () => {
    it('should find the button via testId', () => {
    const testIdName = 'chooseImageButton';
    const {getByTestId} = render(<CreatePost2 />);
    const foundButton = getByTestId(testIdName);
    expect(foundButton).toBeTruthy();
    });

    it('button should say choose image', () => {
        const text = 'Choose image';
        const notFoundText = 'error :(';
        const {toJSON, getByText, queryByText} = render(<CreatePost2 />);
        const foundTextElement = getByText(text);
        const notFoundTextElement = queryByText(notFoundText);
        expect(foundTextElement.props.children).toEqual(text);
        expect(notFoundTextElement).toBeNull();
        expect(toJSON()).toMatchSnapshot();
      });

      it('choose image button is pressable', () => {
        const page = render(<CreatePost2/>);
        const chooseImageButton = page.getByTestId('chooseImageButton');
        fireEvent.press(chooseImageButton);
      })
});

// Testing caption box

const CAPTION_TEXT = 'caption';

describe('Testing caption input', () => {
  it('should find the input box via testId', () => {
  const testIdName = 'captionBox';
  const {getByTestId} = render(<CreatePost2 />);
  const foundButton = getByTestId(testIdName);
  expect(foundButton).toBeTruthy();
  });

    it('Input box is pressable', () => {
      const page = render(<CreatePost2/>);
      const captionBox = page.getByTestId('captionBox');
      fireEvent.press(captionBox);
    });

    test('Input box should allow for input', () => {
      render(<CreatePost2/>);
      fireEvent.changeText(screen.getByPlaceholderText('Write caption...'), CAPTION_TEXT)
    })
});

// Testing location box

const LOCATION_TEXT = 'location';

describe('Testing location input', () => {
  it('should find the location box via testId', () => {
  const testIdName = 'locationBox';
  const {getByTestId} = render(<CreatePost2 />);
  const foundButton = getByTestId(testIdName);
  expect(foundButton).toBeTruthy();
  });

    it('Location box is pressable', () => {
      const page = render(<CreatePost2/>);
      const locationBox = page.getByTestId('locationBox');
      fireEvent.press(locationBox);
    });

    test('Location box should allow for input', () => {
      render(<CreatePost2/>);
      fireEvent.changeText(screen.getByPlaceholderText('Add Location'), LOCATION_TEXT)
    })
});

// Testing "post" button
describe('Testing post button', () => {
    it('should find the button via testId', () => {
    const testIdName = 'postButton';
    const {getByTestId} = render(<CreatePost2 />);
    const foundButton = getByTestId(testIdName);
    expect(foundButton).toBeTruthy();
    });

    it('button should say post', () => {
        const text = 'Post';
        const notFoundText = 'error :(';
        const {toJSON, getByText, queryByText} = render(<CreatePost2 />);
        const foundTextElement = getByText(text);
        const notFoundTextElement = queryByText(notFoundText);
        expect(foundTextElement.props.children).toEqual(text);
        expect(notFoundTextElement).toBeNull();
        expect(toJSON()).toMatchSnapshot();
      });

      it('post button is pressable', () => {
        const navigation = {navigate: () =>{}}
        const page = render(<CreatePost2 navigation={navigation}/>);
        const postButton = page.getByTestId('postButton');
        fireEvent.press(postButton);
      })
      it('Post button navigates to home screen', () => {
            const navigation = {navigate: () =>{}}
            const page = render(<CreatePost2 navigation={navigation}/>);
            jest.spyOn(navigation, 'navigate')
            const postButton = page.getByTestId('postButton');
            fireEvent.press(postButton);
            expect(navigation.navigate).toHaveBeenCalledWith("Home")
          });
});