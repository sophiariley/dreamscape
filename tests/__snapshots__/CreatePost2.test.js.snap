// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing choose image button button should say choose image 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "height": 1334,
            "width": 750,
          }
        }
      >
        <View
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#F6F6F6",
              "backgroundOpacity": 0.8,
              "borderRadius": 5,
              "height": "50%",
              "top": 10,
              "width": "90%",
            }
          }
        />
        <View
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#3A6496",
              "borderRadius": 5,
              "bottom": -15,
              "height": "5%",
              "width": "40%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="chooseImageButton"
          >
            <Text
              style={
                {
                  "alignSelf": "center",
                  "color": "#F6F6F6",
                  "fontSize": 20,
                  "paddingTop": 2,
                }
              }
            >
              Choose image
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#F6F6F6",
              "borderRadius": 5,
              "bottom": -20,
              "height": "20%",
              "opacity": 0.8,
              "width": "90%",
            }
          }
        >
          <TextInput
            onChangeText={[Function]}
            placeholder="Write caption..."
            style={
              {
                "borderRadius": 5,
                "color": "#3A6496",
                "fontSize": 20,
                "height": "20%",
                "marginBottom": 10,
                "marginTop": 5,
                "opacity": 0.5,
                "paddingLeft": 10,
                "width": "80%",
              }
            }
            testID="captionBox"
            value=""
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "bottom": 0,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "alignSelf": "flex-start",
                "flexDirection": "row",
                "left": 10,
                "paddingBottom": 30,
                "paddingTop": 20,
                "position": "absolute",
                "top": 10,
              }
            }
          >
            <TextInput
              onChangeText={[Function]}
              placeholder="Add Location"
              placeholderTextColor="#3A6496"
              style={
                {
                  "color": "#3A6496",
                  "fontSize": 20,
                  "padding": 5,
                }
              }
              testID="locationBox"
              value=""
            />
          </View>
          <View
            style={
              {
                "alignSelf": "flex-end",
                "backgroundColor": "#D28A8E",
                "borderRadius": 5,
                "bottom": -30,
                "height": "27%",
                "right": 20,
                "width": "30%",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="postButton"
            >
              <Text
                style={
                  {
                    "alignSelf": "center",
                    "color": "#F6F6F6",
                    "fontSize": 20,
                    "padding": 12,
                  }
                }
              >
                Post
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Testing post button button should say post 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "height": 1334,
            "width": 750,
          }
        }
      >
        <View
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#F6F6F6",
              "backgroundOpacity": 0.8,
              "borderRadius": 5,
              "height": "50%",
              "top": 10,
              "width": "90%",
            }
          }
        />
        <View
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#3A6496",
              "borderRadius": 5,
              "bottom": -15,
              "height": "5%",
              "width": "40%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="chooseImageButton"
          >
            <Text
              style={
                {
                  "alignSelf": "center",
                  "color": "#F6F6F6",
                  "fontSize": 20,
                  "paddingTop": 2,
                }
              }
            >
              Choose image
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#F6F6F6",
              "borderRadius": 5,
              "bottom": -20,
              "height": "20%",
              "opacity": 0.8,
              "width": "90%",
            }
          }
        >
          <TextInput
            onChangeText={[Function]}
            placeholder="Write caption..."
            style={
              {
                "borderRadius": 5,
                "color": "#3A6496",
                "fontSize": 20,
                "height": "20%",
                "marginBottom": 10,
                "marginTop": 5,
                "opacity": 0.5,
                "paddingLeft": 10,
                "width": "80%",
              }
            }
            testID="captionBox"
            value=""
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "bottom": 0,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "alignSelf": "flex-start",
                "flexDirection": "row",
                "left": 10,
                "paddingBottom": 30,
                "paddingTop": 20,
                "position": "absolute",
                "top": 10,
              }
            }
          >
            <TextInput
              onChangeText={[Function]}
              placeholder="Add Location"
              placeholderTextColor="#3A6496"
              style={
                {
                  "color": "#3A6496",
                  "fontSize": 20,
                  "padding": 5,
                }
              }
              testID="locationBox"
              value=""
            />
          </View>
          <View
            style={
              {
                "alignSelf": "flex-end",
                "backgroundColor": "#D28A8E",
                "borderRadius": 5,
                "bottom": -30,
                "height": "27%",
                "right": 20,
                "width": "30%",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="postButton"
            >
              <Text
                style={
                  {
                    "alignSelf": "center",
                    "color": "#F6F6F6",
                    "fontSize": 20,
                    "padding": 12,
                  }
                }
              >
                Post
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`renders correctly across screens 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "white",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          {
            "height": 1334,
            "width": 750,
          }
        }
      >
        <View
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#F6F6F6",
              "backgroundOpacity": 0.8,
              "borderRadius": 5,
              "height": "50%",
              "top": 10,
              "width": "90%",
            }
          }
        />
        <View
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#3A6496",
              "borderRadius": 5,
              "bottom": -15,
              "height": "5%",
              "width": "40%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="chooseImageButton"
          >
            <Text
              style={
                {
                  "alignSelf": "center",
                  "color": "#F6F6F6",
                  "fontSize": 20,
                  "paddingTop": 2,
                }
              }
            >
              Choose image
            </Text>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={
            {
              "alignSelf": "center",
              "backgroundColor": "#F6F6F6",
              "borderRadius": 5,
              "bottom": -20,
              "height": "20%",
              "opacity": 0.8,
              "width": "90%",
            }
          }
        >
          <TextInput
            onChangeText={[Function]}
            placeholder="Write caption..."
            style={
              {
                "borderRadius": 5,
                "color": "#3A6496",
                "fontSize": 20,
                "height": "20%",
                "marginBottom": 10,
                "marginTop": 5,
                "opacity": 0.5,
                "paddingLeft": 10,
                "width": "80%",
              }
            }
            testID="captionBox"
            value=""
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "bottom": 0,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "alignSelf": "flex-start",
                "flexDirection": "row",
                "left": 10,
                "paddingBottom": 30,
                "paddingTop": 20,
                "position": "absolute",
                "top": 10,
              }
            }
          >
            <TextInput
              onChangeText={[Function]}
              placeholder="Add Location"
              placeholderTextColor="#3A6496"
              style={
                {
                  "color": "#3A6496",
                  "fontSize": 20,
                  "padding": 5,
                }
              }
              testID="locationBox"
              value=""
            />
          </View>
          <View
            style={
              {
                "alignSelf": "flex-end",
                "backgroundColor": "#D28A8E",
                "borderRadius": 5,
                "bottom": -30,
                "height": "27%",
                "right": 20,
                "width": "30%",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
              testID="postButton"
            >
              <Text
                style={
                  {
                    "alignSelf": "center",
                    "color": "#F6F6F6",
                    "fontSize": 20,
                    "padding": 12,
                  }
                }
              >
                Post
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
